### PR TITLE
Parallax fix and Parallax steep occlusion mapping update

### PR DESCRIPTION
**Pull request by scratchyrice**
- Fixed standard/basic parallax mapping as previously was not working and "flat".
- Added steep parallax occlusion mapping
- Added new "parallax_steep" parameter for use in materials
- Added new SteepParallaxOcclusionMapping sample material
- Updated Bump mapping sample. SteepParallaxOcclusionMapping is now selectable. Added cube.mesh to this sample as it shows it better.
- Added Steep parallax occlusion to terrain SRS
- Added seperate Steep parallax mapping option to terrain (setLayerSteepParallaxMappingEnabled to material generator A).
- Updated terrain sample to include steep parallax occlusion. Added a checkbox to sample to turn this on (Off by default).

### CODE REVIEW

File: Components/RTShaderSystem/src/OgreShaderExNormalMapLighting.cpp [hunk 1]
```diff
@@ -123,7 +146,7 @@ bool NormalMapLighting::createCpuSubPrograms(ProgramSet* programSet)
         vstage.assign(In(vsInTangent).w(), Out(vsOutTangent).w());
 
         // transform normal
-        fstage.callFunction("SGX_TransformNormal", {In(viewNormal), In(psInTangent), InOut(newViewNormal)});
+        fstage.callFunction("SGX_TransformNormal", {In(psOutTBN), InOut(newViewNormal)});
```
Comment (paroj) [inline Components/RTShaderSystem/src/OgreShaderExNormalMapLighting.cpp:149]:
```suggestion
        fstage.callBuiltin("mul", psOutTBN,newViewNormal, newViewNormal);
```

File: Media/RTShaderLib/SGXLib_NormalMap.glsl [hunk 2]
```diff
@@ -41,20 +41,198 @@ void SGX_FetchNormal(in sampler2D s,
...
+		float afterDepth  = currentDepthMapValue - currentLayerDepth;
+		float beforeDepth = ((1.0f - texture2D(normalHeightMap, prevTexCoords).a) - scaleBias.y) - currentLayerDepth + layerDepth;
+		 
+		// interpolation of texture coordinates
+		float weight = afterDepth / (afterDepth - beforeDepth);
+		newTexCoord = prevTexCoords * weight + newTexCoord * (1.0 - weight);
+	}
+}
+
+void SGX_Generate_Parallax_Texcoord_Old(in sampler2D normalHeightMap,
```
Comment (paroj) [inline Media/RTShaderLib/SGXLib_NormalMap.glsl:235]: if it is unused, you can just delete it

File: Media/RTShaderLib/SGXLib_NormalMap.glsl [hunk 3]
```diff
@@ -41,20 +41,198 @@ void SGX_FetchNormal(in sampler2D s,
...
 void SGX_Generate_Parallax_Texcoord(in sampler2D normalHeightMap,
+						in vec2 texCoord,
+						in vec3 viewPos,
+						in vec2 scaleBias,
+						in mat3 TBN,
+						out vec2 newTexCoord)
+{
+	//Calculate eye direction
+	TBN = transpose(TBN);
+	vec3 eyeVec = mul(TBN, -viewPos);
```
Comment (paroj) [inline Media/RTShaderLib/SGXLib_NormalMap.glsl:92]:
```suggestion
	vec3 eyeVec = mul(-viewPos, TBN); // transpose(TBN) * -viewPos
```

File: Media/RTShaderLib/SGXLib_NormalMap.glsl [hunk 4]
```diff
@@ -41,20 +41,198 @@ void SGX_FetchNormal(in sampler2D s,
...
+	eyeVec = normalize(eyeVec);
+	
+	//Simple parallax mapping
+	newTexCoord = texCoord;
+	float height = 1.0f - texture2D(normalHeightMap, newTexCoord).a;   
+    vec2 p = eyeVec.xy * (height * scaleBias.x);
+    newTexCoord = newTexCoord - p;
+}
+
+void SGX_Generate_Parallax_Steep_Texcoord_Terrain(in sampler2D normalHeightMap,
```
Comment (paroj) [inline Media/RTShaderLib/SGXLib_NormalMap.glsl:176]: you could `#ifdef TERRAIN_SURFACE` the differences between this and the other version into one.
One will ever need only one variant of this per generated shader.

**Comment by paroj** [comment]
you can consider some of my suggestions if you are updating this anyway. Otherwise I will fix those myself, when I get to properly testing this.
Note, that you can just add commits to this branch - no need to create a new pull-request.

File: Media/RTShaderLib/SGXLib_NormalMap.glsl [hunk 5]
```diff
@@ -41,20 +41,198 @@ void SGX_FetchNormal(in sampler2D s,
...
+    float layerDepth = 1.0 / numLayers;	
+    float currentLayerDepth = 0.0;	
+    vec2 parallaxShift = (eyeVec.xy) * scaleBias.x; 
+    vec2 deltaTexCoords = parallaxShift / numLayers;
+	
+	newTexCoord = texCoord;	
+	float currentDepthMapValue = 1.0f - texture2D(normalHeightMap, newTexCoord).a;
+	
+	//Loop through layers and break early if match found.
+	for (int currentLayerId = 0; currentLayerId < (int)numLayers; currentLayerId++)
```
Comment (paroj) [inline Media/RTShaderLib/SGXLib_NormalMap.glsl:128]:
```suggestion
	for (int currentLayerId = 0; currentLayerId < int(numLayers); currentLayerId++)
```
GLSL does not like the C style cast

Comment (scratchyrice) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl:128]: Thanks Paroj, I will implement these changes and get committed to you today.

Also have another change i want to implement.

File: Components/RTShaderSystem/src/OgreShaderExNormalMapLighting.cpp (hunk 1 above)
Comment (scratchyrice) [reply Components/RTShaderSystem/src/OgreShaderExNormalMapLighting.cpp:149]: Done this.

File: Media/RTShaderLib/SGXLib_NormalMap.glsl (hunk 2 above)
Comment (scratchyrice) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl:235]: Ignore previous - this function has been removed.

File: Media/RTShaderLib/SGXLib_NormalMap.glsl (hunk 3 above)
Comment (scratchyrice) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl:92]: I tried this but unfortunately it breaks the parallax effect - it looks very flat with this change.

Comment (scratchyrice) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl:92]: Oh no sorry - it does work - didn't remove TBN = transpose.

File: Media/RTShaderLib/SGXLib_NormalMap.glsl (hunk 5 above)
Comment (scratchyrice) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl:128]: This fix is done.

File: Media/RTShaderLib/SGXLib_NormalMap.glsl (hunk 4 above)
Comment (scratchyrice) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl:176]: I can't seem to get this to work properly - it just ignores the #ifdef within SGXLib_NormalMap.glsl?

Comment (paroj) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl:176]: ah, sorry.. there is no `TERRAIN_SURFACE` define, I meant you define it.
But there already is `TERRAIN_STEEP_PARALLAX_MAPPING`, so you can use that..

Comment (scratchyrice) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl:176]: > ah, sorry.. there is no `TERRAIN_SURFACE` define, I meant you define it. But there already is `TERRAIN_STEEP_PARALLAX_MAPPING`, so you can use that..

Ah yes I see. I have sorted this now.

File: Media/RTShaderLib/SGXLib_NormalMap.glsl (hunk 5 above)
Comment (scratchyrice) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl]: All done/pushed.

**Comment by scratchyrice** [comment]
Seems its failed on white space!? Best way to fix?

**Comment by scratchyrice** [comment]
@paroj Did you want me to include the new textures I was using which shows it off more? Idk what the ogre rules are regarding media.

**Comment by paroj** [comment]
> @paroj Did you want me to include the new textures I was using which shows it off more? Idk what the ogre rules are regarding media.

the textures must be freely redistributable. Ideally you would take one from the site referenced here (CC0):
https://github.com/OGRECave/ogre/blob/master/Samples/Media/materials/textures/terrain/README.md

Also I downscaled them to 1024px as being small is more important than visual fidelity for the samples

File: Media/RTShaderLib/SGXLib_NormalMap.glsl [hunk 6]
```diff
@@ -58,11 +84,79 @@ void SGX_Generate_Parallax_Texcoord(in sampler2D normalHeightMap,
...
+
+    newTexCoord = newTexCoord - p;
+}
+
+void SGX_Generate_Parallax_Steep_Texcoord(in sampler2D normalHeightMap,
+						in vec2 texCoord,
+						in vec3 viewPos,
+						in vec2 scaleBias,
+						in float layerCount,
+						in float maxDistance,
```
Comment (paroj) [inline Media/RTShaderLib/SGXLib_NormalMap.glsl:113]: this parameter is no longer used now

**Comment by scratchyrice** [comment]
> > @paroj Did you want me to include the new textures I was using which shows it off more? Idk what the ogre rules are regarding media.
> 
> the textures must be freely redistributable. Ideally you would take one from the site referenced here (CC0): https://github.com/OGRECave/ogre/blob/master/Samples/Media/materials/textures/terrain/README.md
> 
> Also I downscaled them to 1024px as being small is more important than visual fidelity for the samples

Okay cool - thats the site I use anyway. Ill take a look and see what shows the effect up best.

File: Media/RTShaderLib/SGXLib_NormalMap.glsl (hunk 6 above)
Comment (scratchyrice) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl:113]: Ooops! Fixed.

**Comment by scratchyrice** [comment]
> > @paroj Did you want me to include the new textures I was using which shows it off more? Idk what the ogre rules are regarding media.
> 
> the textures must be freely redistributable. Ideally you would take one from the site referenced here (CC0): https://github.com/OGRECave/ogre/blob/master/Samples/Media/materials/textures/terrain/README.md
> 
> Also I downscaled them to 1024px as being small is more important than visual fidelity for the samples

OGRE Sample Browser 27_06_2023 16_37_58

OGRE Sample Browser 27_06_2023 16_38_01

OGRE Sample Browser 27_06_2023 16_38_10

Look good to you paroj? They're 1024x1024 dds textures.

File: Media/RTShaderLib/SGXLib_NormalMap.glsl (hunk 6 above)
Comment (paroj) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl:113]: I did not even notice `scaleBias`.. I was referring to `maxDistance`

Comment (scratchyrice) [reply Media/RTShaderLib/SGXLib_NormalMap.glsl:113]: > I did not even notice `scaleBias`.. I was referring to `maxDistance`

Ah! Sorry, it should be used - the conditional was removed somehow. Framerate plummets for no good reason with an infinate distance as you cant see it from a distance anyway. Ill add this back in.

**Comment by paroj** [comment]
- renamed this to parallax occlusion mapping, as all other engines call it like that
- added specular to the new texture, so it can be used with terrain and kept the ambientCG name

**Pull request merged**